### ISSUE TITLE
Convince eldev to not bootstrap?

### ISSUE THREAD

**Issue by sellout** [doublep/eldev#81]
I’m handing eldev an Emacs with pinned versions of all the packages (eldev, relint, etc.) already there on the load-path. However, eldev still wants to bootstrap and pull everything down itself. How can I convince it to use what’s already provided?

Options like `--external` only seem to apply to dependencies of my project (which makes sense), but I’m hoping there’s something similar for _eldev’s_ dependencies.

**Comment by sellout** [comment]
To clarify a bit, I suppose this isn’t quite a bootstrapping issue, as I’m using `ELDEV_LOCAL` to point to the local eldev, which works. And what I thought _wasn’t_ working, but _is_ is using `(eldev-use-local-dependency "path/to/dash")` etc. to specify the other dependencies.

However, it still wants to download package archives, even though it doesn’t need to load anything from them, and that’s where I’m currently stuck. I tried `(setq package-archives nil)` in my Eldev, but that didn’t have any effect.

I finally got it. Four things I needed to do:
1. `ELDEV_LOCAL` to avoid downloading Eldev,
2. `eldev-use-local-dependency` for each linter & dependency needed,
3. `eldev --external` to avoid downloading any archives, and
4. `export HOME="$PWD/fake-home` so Eldev could write its pseudo-package temp file somewhere.

This is still a bit awkward in a couple was (e.g., step 2 requires explicitly listing transitive dependencies), but I have a successful build now that doesn’t try to do any networking.

**Comment by doublep** [comment]
Yeah, this sounds like a bug. I guess Eldev should use provided tools in `--external` mode if they are enough. I'm not sure if it should try fetching if something is e.g. not installed (does that even work?).

Maybe as a workaround you could try modifying `eldev-known-tool-packages` for now, by stripping `:archive`/`:archives'. But I will look into fixing this eventually.

**Comment by doublep** [comment]
I now tried to reproduce it, but couldn't. I don't have `package-lint` installed normally, so what I did is trying to run the linter in external mode from some directory:

    ~/iter2$ eldev clean all; eldev clean global-cache; eldev -X -dt lint package
    Deleted 1 directory
    Nothing to delete
    Started up on Sat Jan 21 13:58:43 2023
    Running on GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 2.24.33, cairo version 1.16.0)
     of 2022-12-03
    Project directory: ‘/home/paul/iter2/’
    Loading file ‘/home/paul/.eldev/config’...
    Loading file ‘Eldev’...
    Loading file ‘Eldev-local’...
    Executing command ‘lint’...
    Going to run the following linter: ‘package’
    Running linter ‘package’...
    Need the following extra dependencies for set ‘runtime’: package-lint 0.14

    Dependency ‘package-lint’ version 0.14 is required, but only 0.12 is installed in ‘~/.emacs.d/elpa’; skipping linter ‘package’
    Linter has no complaints
    Deleting package archive contents to avoid polluting the external directory: ‘--eldev--’
    Finished successfully on Sat Jan 21 13:58:44 2023

It never accesses the network. I even double-verified this by termporarily disabling network connection on my machine. The message it gives is not exactly correct, but that's it. I already have a fix and improvement locally, after which it says instead:

    Dependency ‘package-lint’ is not installed in ‘~/.emacs.d/elpa’; skipping linter ‘package’
    Either install it there, or use global option ‘--isolated’ (‘-I’)
    Required as a development tool, not as a dependecy of the project

However, the only network access I could reproduce is bootstrapping Eldev *itself* if not available locally (i.e. after setting `ELDEV_LOCAL` to an empty string and making sure it is not installed yet). However, that is understandable, as if Eldev package is missing, Eldev obviously cannot run. Also note that `-X` (`--external`) doesn't apply here, since the package is needed before even parsing the command line, so it needs to be present in the standard location regardless of `-X`.

I need a full description of the setup where it misbehaves for you.

**Comment by sellout** [comment]
An update a couple years later …

The addition of `--disable-dependencies` has been very helpful. It allowed me to remove the transitive application of `eldev-use-local-dependency` that I needed with `--external` (I’m working in Nix, so it’s very similar to the Guix situation that inspired that flag).

However, something else changed that I’m currently trying to figure out. Even though I have `ELDEV_LOCAL` set to a path that seems to be correct, Eldev always wants to bootstrap itself. I couldn’t find any code that had changed on either my side or in Eldev that seemed suspicious.

One project that I’m using this in is [auto-dark-emacs](). [Here is my latest code](https://github.com/sellout/auto-dark-emacs/blob/fix-nix/nix/packages/auto-dark.nix#L24-L47) (although I don’t expect the Nix to be obvious). That snippet catches the use of `ELDEV_LOCAL`, setting a fake `HOME`, and using `--disable-dependencies`. But the output when it gets to the Eldev call is
```
Bootstrapping Eldev for Emacs 30.1 from MELPA Stable...
Error: error ("stable.melpa.org/80 Temporary failure in name resolution (when updating contents of package archive `melpa-stable')")
```

Interestingly, when I do the same stuff for [`doctor`](https://github.com/sellout/auto-dark-emacs/blob/5d5baf89be2c4aa3c305c636d03a93c0154cd985/nix/checks.nix#L11-L28) and [`lint`](https://github.com/sellout/auto-dark-emacs/blob/5d5baf89be2c4aa3c305c636d03a93c0154cd985/nix/checks.nix#L48-L64), I have no trouble.

It would be nice if Eldev reported why it’s not using the `ELDEV_LOCAL` setting (e.g., if one of the `package-desc-reqs` is missing … but the only one of those I see is `emacs`, which is trivially satisfied).

**Comment by doublep** [comment]
Bootstrapping happens not from the main program, but from the script `bin/eldev` if it cannot activate the main package (including from `ELDEV_LOCAL`). I suggest you try to debug it yourself. Since there we don't have access to command line options yet, we cannot add something like `(eldev-verbose ...)` etc., so all output from the script is unconditional — that's why I don't want to add more output than the only "Bootstrapping...." message.
